polaris-migrator/src/utilities/sass.ts

```diff
@@ -251,3 +263,398 @@ export function createInlineComment(text: string, options?: {prose?: boolean}) {
+      report: (report: Report) => void;
+      each: <T extends Container>(
+        root: T,
+        walker: (decl: PostCSSNode) => false | void,
+      ) => void;
+      walk: <T extends Container>(
+        root: T,
+        walker: (decl: PostCSSNode) => false | void,
+      ) => void;
+      walkComments: <T extends Container>(
+        root: T,
+        walker: (comment: PostCSSComment) => false | void,
+      ) => void;
+      walkAtRules: <T extends Container>(
+        root: T,
+        atRuleWalker: (
+          atRule: AtRule,
+          parsedValue: ParsedValue,
+        ) => false | void,
+      ) => void;
+      walkDecls: <T extends Container>(
+        root: T,
+        declWalker: (
+          decl: Declaration,
+          parsedValue: ParsedValue,
+        ) => false | void,
+      ) => void;
+      walkRules: <T extends Container>(
+        root: T,
+        ruleWalker: (decl: PostCSSRule) => false | void,
+      ) => void;
+    };
+  },
+  context: PluginContext,
+) => (root: Root) => void;
+
+// Expose a stylelint-like API for creating sass migrators so we can easily
+// migrate to that tool in the future.
+function convertStylelintRuleToPostcssProcessor(ruleFn: StylelintRule) {
+  return (fileInfo: FileInfo, _: API, options: Options) => {
+    const plugin: Plugin = {
+      postcssPlugin: ruleFn.ruleName,
+      // PostCSS will rewalk the AST every time a declaration/rule/etc is
+      // mutated by a plugin. This can be useful in some cases, but in ours we
+      // only want single-pass behaviour.
+      //
+      // This can be avoided in 2 ways:
+      //
+      // 1) Flagging each declaration as we pass it, then skipping it on
+      //    subsequent passes.
+      // 2) Using postcss's Once() plugin callback.
+      //
+      // stylelint also uses `Once()`, so we're able to remove this once we've
+      // migrated:
+      // https://github.com/stylelint/stylelint/blob/cb425cb/lib/postcssPlugin.js#L22
+      Once(root: Root) {
+        // NOTE: For fullest compatibility with stylelint, we initialise the
+        // rule here _inside_ the postcss Once function just like stylelint
+        // does. I _think_ this means multiple passes can be performed without
+        // rules accidentally retaining scoped variables, etc.
+        const ruleProcessor = ruleFn(
+          // Normally, this comes from stylelint config, but for this shim we
+          // just hard-code it, and instead rely on the "seconary" options
+          // object for passing through the jscodeshift options.
+          true,
+          {
+            ...options,
+            partialFixStrategy: 'report',
+            __injectReportsAsComments: true,
+          },
+          {fix: true},
+        );
+
+        ruleProcessor(root);
+      },
+    };
+
+    return postcss(plugin).process(fileInfo.source, {
+      syntax: require('postcss-scss'),
+    }).css;
+  };
+}
+
+export function createStylelintRule(ruleName: string, ruleFn: PolarisMigrator) {
```

Comment:
This factory function matches the `stylelint` API as closely as possible. There are only a couple of lines in this function which would be removed once we move to `stylelint`.